clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3442,147 +3488,132 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
-            } else {
-              SYCLKernel->addAttr(A);
-            }
-          } else {
-            SYCLKernel->addAttr(A);
-          }
-          break;
-        }
-        case attr::Kind::SYCLIntelMaxWorkGroupSize: {
-          auto *SIMWGSA = cast<SYCLIntelMaxWorkGroupSizeAttr>(A);
-          if (auto *Existing = SYCLKernel->getAttr<ReqdWorkGroupSizeAttr>()) {
-            ASTContext &Ctx = getASTContext();
-            if (Existing->getXDimVal(Ctx) > SIMWGSA->getXDimVal(Ctx) ||
-                Existing->getYDimVal(Ctx) > SIMWGSA->getYDimVal(Ctx) ||
-                Existing->getZDimVal(Ctx) > SIMWGSA->getZDimVal(Ctx)) {
-              Diag(SYCLKernel->getLocation(),
-                   diag::err_conflicting_sycl_kernel_attributes);
-              Diag(Existing->getLocation(), diag::note_conflicting_attribute);
-              Diag(SIMWGSA->getLocation(), diag::note_conflicting_attribute);
-              SYCLKernel->setInvalidDecl();
-            } else {
-              SYCLKernel->addAttr(A);
-            }
-          } else {
-            SYCLKernel->addAttr(A);
-          }
-          break;
-        }
-        case attr::Kind::SYCLIntelKernelArgsRestrict:
-        case attr::Kind::SYCLIntelNumSimdWorkItems:
-        case attr::Kind::SYCLIntelSchedulerTargetFmaxMhz:
-        case attr::Kind::SYCLIntelMaxGlobalWorkDim:
-        case attr::Kind::SYCLIntelNoGlobalWorkOffset:
-        case attr::Kind::SYCLIntelUseStallEnableClusters:
-        case attr::Kind::SYCLIntelLoopFuse:
-        case attr::Kind::SYCLIntelFPGAMaxConcurrency:
-        case attr::Kind::SYCLIntelFPGADisableLoopPipelining:
-        case attr::Kind::SYCLIntelFPGAInitiationInterval:
-        case attr::Kind::SYCLSimd: {
-          if ((A->getKind() == attr::Kind::SYCLSimd) && KernelBody &&
-              !KernelBody->getAttr<SYCLSimdAttr>()) {
-            // Usual kernel can't call ESIMD functions.
-            Diag(KernelBody->getLocation(),
-                 diag::err_sycl_function_attribute_mismatch)
-                << A;
-            Diag(A->getLocation(), diag::note_attribute);
-            KernelBody->setInvalidDecl();
-          } else
-            SYCLKernel->addAttr(A);
-          break;
-        }
-        // TODO: vec_len_hint should be handled here
-        default:
-          // Seeing this means that CollectPossibleKernelAttributes was
-          // updated while this switch wasn't...or something went wrong
-          llvm_unreachable("Unexpected attribute was collected by "
-                           "CollectPossibleKernelAttributes");
-        }
+  case attr::Kind::ReqdWorkGroupSize: {
+    auto *RWGSA = cast<ReqdWorkGroupSizeAttr>(A);
+    if (auto *Existing = SYCLKernel->getAttr<ReqdWorkGroupSizeAttr>()) {
+      ASTContext &Ctx = S.getASTContext();
+      if (Existing->getXDimVal(Ctx) != RWGSA->getXDimVal(Ctx) ||
+          Existing->getYDimVal(Ctx) != RWGSA->getYDimVal(Ctx) ||
+          Existing->getZDimVal(Ctx) != RWGSA->getZDimVal(Ctx)) {
```

Comment:
This isn't specific to your review, but I think there's a bug here -- we're not checking to see how this attribute is spelled and so we may be comparing things we don't expect to compare. e.g., the OpenCL spelling for this attribute has argument order X, Y, Z and the SYCL spelling has argument order Z, Y, X.

Reply:
lol... that is depressing.  I just moved this code though.

Are we sure that the SemaDeclAttr part doesn't correct the order of these?

Reply:
Yeah, that's why this is an existing bug. I'm pretttttty sure that SemaDeclAttr isn't correcting the order of those because we need to keep the order sensible for AST dumping, pretty printing, etc. IIRC, we tried that route but had to roll it back. @smanna12 may know more (or be able to look into whether this is a real issue or just me being paranoid).

Reply:
Yes, this is a bug with the existing implementation. I plan to refactor this attribute.